### PR TITLE
HS-288: Notification Service Alarm Kafka Consumer

### PR DESCRIPTION
**Pull request by thomasbigger584**
## Description
Adding a Kafka Consumer to consume both AlarmDTO from a topic in the kafka instance.

Currently when we run `skaffold dev --trigger=manual` the notification service comes up before Kafka. There is some output that the notification service cannot connect to Kafka but then _does_ eventually connect when Kafka is up, around ~30 seconds in.

@geraldhumphries Do you know of a way for the notification service to depend on the Kafka deployment being up. I was thinking similar to docker-compose `depends_on`, and some google searches point towards `initContainers`: https://kubernetes.io/docs/concepts/workloads/pods/init-containers/, https://stackoverflow.com/a/52705936, https://stackoverflow.com/a/72345864 - what is your thoughts on this?

Edit: have added an `initContainer` which helps it a bit, although it still needs to wait for a leader to become available. 

## Jira link(s)
- https://issues.opennms.org/browse/HS-288

### CODE REVIEW

**Review by james8hut — COMMENTED** [review]
Looks good to me

File: notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import org.opennms.horizon.shared.dto.event.AlarmDTO;
+import org.opennms.horizon.shared.dto.event.EventDTO;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Service;
+
+@Service
+public class KafkaConsumer {
+    private final Logger log = LoggerFactory.getLogger(KafkaConsumer.class);
```
Comment (cgorantla) [inline notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java:12]: Just for consistency, that's how most of opennms code uses LOG
```suggestion
    private final Logger LOG = LoggerFactory.getLogger(KafkaConsumer.class);
```

File: notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+    @KafkaListener(
+        topics = "${horizon.kafka.events.topic}",
+        concurrency = "${horizon.kafka.events.concurrency}"
+    )
+    public void consume(EventDTO event) {
+        log.info(String.format("Consumed Event: %s", event.getId()));
```
Comment (cgorantla) [inline notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java:19]: may be use {} pattern with slf4j logger

File: notifications/src/main/java/org/opennms/horizon/notifications/rest/NotificationRestController.java [hunk 3]
```diff
@@ -43,9 +48,30 @@ public class NotificationRestController {
...
     @GetMapping("/pagerDutyKey")
     public ResponseEntity<String> getPagerDutyKey() {
         String key = notificationsService.getPagerDutyKey();
         return new ResponseEntity<>(key, HttpStatus.OK);
     }
+
+    /*
+     * Added only for test purposes
+     */
+    @PostMapping("/publishKafkaEvent")
```
Comment (cgorantla) [inline notifications/src/main/java/org/opennms/horizon/notifications/rest/NotificationRestController.java:63]: Can be removed and KafkaProducer can be moved to test.  It should be simple enough to use kafka producer, may not need Rest call.

File: notifications/src/main/resources/application.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+spring:
+  application:
+    name: horizon-stream-notifications
+
+  kafka:
+    consumer:
+      group-id: org.opennms.horizon.notifications
+      auto-offset-reset: earliest
+      key-deserializer: org.apache.kafka.common.serialization.StringDeserializer
+      value-deserializer: org.springframework.kafka.support.serializer.JsonDeserializer
```
Comment (cgorantla) [inline notifications/src/main/resources/application.yml:10]: Not entirely sure if this will remain Json but good for now.

**Comment by cgorantla** [comment]
Happened to see this post https://medium.com/@kelseyhightower/12-fractured-apps-1080c73d481c.  If possible avoid depending on Kafka being deployed before the notification service listener starts.

**Comment by geraldhumphries** [comment]
> Happened to see this post https://medium.com/@kelseyhightower/12-fractured-apps-1080c73d481c. If possible avoid depending on Kafka being deployed before the notification service listener starts.

@cgorantla I took a look at that post, and noticed that it's from 2015, which is pre-Kuberentes. K8s enables the "start-crash-restart" pattern when it comes to service startup order. It essentially externalizes "do connection retries in the application" as the article suggests. 

>Configuration management tools can solve the “missing” database problem by ensuring the database is started before the applications that depend on it. This is nothing more then a band-aid covering up the larger problem. The application should simply retry the database connection, using some sort of backoff, and log errors along the way. At some point either the database will come online, or your company will be out of business.

If we do require more resilient pods, then we SHOULD follow that suggestion. I don't think we need that right now.

**Comment by cgorantla** [comment]
> > Happened to see this post https://medium.com/@kelseyhightower/12-fractured-apps-1080c73d481c. If possible avoid depending on Kafka being deployed before the notification service listener starts.
> 
> @cgorantla I took a look at that post, and noticed that it's from 2015, which is pre-Kuberentes. K8s enables the "start-crash-restart" pattern when it comes to service startup order. It essentially externalizes "do connection retries in the application" as the article suggests.
> 
> > Configuration management tools can solve the “missing” database problem by ensuring the database is started before the applications that depend on it. This is nothing more then a band-aid covering up the larger problem. The application should simply retry the database connection, using some sort of backoff, and log errors along the way. At some point either the database will come online, or your company will be out of business.
> 
> If we do require more resilient pods, then we SHOULD follow that suggestion. I don't think we need that right now.

Irrespective of who manages the service startup, I think it is possible for notification service not to depend on active kafka connection and starts service.  Consumer poll will timeout and try again.  It's very common practice in Kafka consumer like [here](https://github.com/OpenNMS/opennms/blob/bec6f0ce6cbb1e4eac5903cc1004619c1ff60804/core/ipc/sink/kafka/server/src/main/java/org/opennms/core/ipc/sink/kafka/server/KafkaMessageConsumerManager.java#L163) as far as i understand.

**Comment by thomasbigger584** [comment]
> > > Happened to see this post https://medium.com/@kelseyhightower/12-fractured-apps-1080c73d481c. If possible avoid depending on Kafka being deployed before the notification service listener starts.
> > 
> > 
> > @cgorantla I took a look at that post, and noticed that it's from 2015, which is pre-Kuberentes. K8s enables the "start-crash-restart" pattern when it comes to service startup order. It essentially externalizes "do connection retries in the application" as the article suggests.
> > > Configuration management tools can solve the “missing” database problem by ensuring the database is started before the applications that depend on it. This is nothing more then a band-aid covering up the larger problem. The application should simply retry the database connection, using some sort of backoff, and log errors along the way. At some point either the database will come online, or your company will be out of business.
> > 
> > 
> > If we do require more resilient pods, then we SHOULD follow that suggestion. I don't think we need that right now.
> 
> Irrespective of who manages the service startup, I think it is possible for notification service not to depend on active kafka connection and starts service. Consumer poll will timeout and try again. It's very common practice in Kafka consumer like [here](https://github.com/OpenNMS/opennms/blob/bec6f0ce6cbb1e4eac5903cc1004619c1ff60804/core/ipc/sink/kafka/server/src/main/java/org/opennms/core/ipc/sink/kafka/server/KafkaMessageConsumerManager.java#L163) as far as i understand.

Thanks for the clarification, sounds like we want the application to retry the kafka connection on startup, as opposed to waiting for it to be up.

**Comment by thomasbigger584** [comment]
Changes made:
- Removal of `initContainer` in favor of having the application retry connection failures on startup
- Removal of consumption from an events topic as it is not required.
- Removal of unused testing endpoint for producer and tested using an external producer(https://kafka.apache.org/downloads)
1. It requires a change in `/etc/hosts` in order for the connection to the broker to work:
```
$ sudo vim /etc/hosts
```
2. Edit the localhost line to include `kafka`:
```
127.0.0.1       localhost kafka
```
3. Use the producer script to test sending a message:
```
$ ./kafka-console-producer.sh --topic org.opennms.horizon.notifications.alarms  --bootstrap-server kafka:9092
> {"id": 12345}
```
4. Message gets consumed by the configured `@KafkaListener`

File: notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java (hunk 1 above)
Comment (thomasbigger584) [reply notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java]: Do we prefer this line or Lomboks `@Slf4j` annotation?

Comment (cgorantla) [reply notifications/src/main/java/org/opennms/horizon/notifications/kafka/KafkaConsumer.java]: I'm not fan of Lomboks. You can keep this.

**Comment by cgorantla** [comment]
> Changes made:
> 
> * Removal of `initContainer` in favor of having the application retry connection failures on startup.
> * Removal of consumption from an events topic as it is not required.
> * Removal of unused testing endpoint for producer and tested using an external producer(https://kafka.apache.org/downloads)
> 
> 1. It requires a change in `/etc/hosts` in order for the connection to the broker to work for local testing with skaffold:
> 
> ```
> $ sudo vim /etc/hosts
> ```
> 
> 2. Edit the localhost line to include `kafka`:
> 
> ```
> 127.0.0.1       localhost kafka
> ```
> 
> 3. Use the producer script to test sending a message:
> 
> ```
> $ ./kafka-console-producer.sh --topic org.opennms.horizon.notifications.alarms  --bootstrap-server kafka:9092
> > {"id": 12345}
> ```
> 
> 4. Message gets consumed by the configured `@KafkaListener`

You should be able to expose local port for Kafka in skaffold.yaml for testing https://github.com/OpenNMS/horizon-stream/blob/ccff977997e2fb5b4eeef83f048408445c4a7fa5/skaffold.yaml#L111

**Comment by thomasbigger584** [comment]
> > Changes made:
> > 
> > * Removal of `initContainer` in favor of having the application retry connection failures on startup.
> > * Removal of consumption from an events topic as it is not required.
> > * Removal of unused testing endpoint for producer and tested using an external producer(https://kafka.apache.org/downloads)
> > 
> > 
> > 1. It requires a change in `/etc/hosts` in order for the connection to the broker to work for local testing with skaffold:
> > 
> > ```
> > $ sudo vim /etc/hosts
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 2. Edit the localhost line to include `kafka`:
> > 
> > ```
> > 127.0.0.1       localhost kafka
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 3. Use the producer script to test sending a message:
> > 
> > ```
> > $ ./kafka-console-producer.sh --topic org.opennms.horizon.notifications.alarms  --bootstrap-server kafka:9092
> > > {"id": 12345}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > 4. Message gets consumed by the configured `@KafkaListener`
> 
> You should be able to expose local port for Kafka in skaffold.yaml for testing
> 
> https://github.com/OpenNMS/horizon-stream/blob/ccff977997e2fb5b4eeef83f048408445c4a7fa5/skaffold.yaml#L111

Yeah this has been added, although due to this configuration https://github.com/OpenNMS/horizon-stream/blob/develop/dev/kubernetes.kafka.yaml#L431-L432 external brokers will timeout and not connect. I think @geraldhumphries is aware of this.

**Comment by thomasbigger584** [comment]
Changes made working around the need to edit `/etc/hosts`. 

Using a localclient and exposing that on another port. NotificationService connects to 9092 as usual, but local producers through skaffold connect on port 9090.

**Comment by thomasbigger584** [comment]
By reading this article (https://www.confluent.io/en-gb/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/#adding-new-listener) it would imply that having 2 listeners, `kafka:9092` for communication internally with other pods, and `localhost:59092` for using a local producer outside of the cluster, they need to be configured to use different ports.
This picture shows the current configuration: https://cdn.confluent.io/wp-content/uploads/listeners1.018-768x543.png

If anyone knows of another way to get localhost connected to the broker, please let me know.

**Review by cgorantla — APPROVED** [review]
LGTM!

I'm not sure if we need to expose kafka outside the cluster. I think you are doing it to test  Kafka alarms but that may not be needed in future when we have kafka producer that is sending alarms to kafka.

**Pull request merged**